GHzDACs/jump_table.py

```diff
@@ -0,0 +1,629 @@
+            c = list(counters)
+            while len(c) < cls.NUM_COUNTERS:
+                c.append(0)
+        return c
+
+    def __str__(self):
+        counter = '\n'.join(
+            "Counter {}: {}".format(
+                i, c
+            ) for i, c in enumerate(self.counters)
+        )
+        start_addr = 'Start address: %d' % self.start_addr
+        jump = '\n'.join(
+            "-JUMP ENTRY {}-\n{}".format(
+                i, str(jump)
+            ) for i, jump in enumerate(self.jumps)
+        )
+        return '\n'.join([counter, start_addr, jump])
+
+    def toString(self):
+        """Serialize jump table to a byte string for the FPGA"""
+        data = np.zeros(self.PACKET_LEN, dtype='<u1')
+        # Set counter values. Each one is 4 bytes
+        for i, c in enumerate(self.counters):
+            data[i * 4:(i + 1) * 4] = littleEndian(c, 4)
+        # Set start address
+        data[16:19] = littleEndian(self.start_addr, 3)
+        data[19:22] = littleEndian(self.start_addr, 3)
+        # Start op code
+        data[22] = 5
+        data[23] = 0
+        for i, jump in enumerate(self.jumps):
+            ofs = 24 + i * 8
+            data[ofs:ofs+8] = jump.as_bytes()
+        return data.tostring()
+
+    def pretty_string(self):
+        s = ''
+        data = self.toString()
+        i = 0
+        # counters
+        for j in range(4):
+            s += '{0:02x} {1:02x} {2:02x} {3:02x}'.format(*[ord(x) for x in data[i:i+4]])
+            i += 4
+            s += '\n'
+        # table commands
+        for j in range((528-16)//8):
+            s += '{0:02x} {1:02x} {2:02x}  '.format(*[ord(x) for x in data[i:i+3]])
+            i += 3
+            s += '{0:02x} {1:02x} {2:02x}  '.format(*[ord(x) for x in data[i:i+3]])
+            i += 3
+            s += '{0:02x} {1:02x}  '.format(*[ord(x) for x in data[i:i+2]])
+            i += 2
+            s += '\n'
+        # print i
+        # print len(data)
+        assert(i == len(data))
+        return s
+
+
+# TEST FUNCTIONS
+
+def testNormal(stopAddr):
+    """Test END function
+    
+    The SRAM block we use to store data
+    
+    ns  || 0         10        20        30
+        || 0123456789012345678901234567890123456789
+    cell|| 0  |1  |2  |3  |4  |5  |6  |7  |8  |9  |
+    data|| ____--------________----____________________
+    table                          End
+    
+    If stopAddr >= 6:
+    The sequence should show a pulse of 8ns length, followed by a gap of 8ns,
+    and then another 4ns pulse. Since the system idles at the same cell as the
+    END command the idle state should be all zero.
+    
+    If stopAddr == 5
+    The system will idle inside the final 4ns pulse, meaning the idle values
+    are high, not zero. The second pulse will not appear to have an end because
+    of this.1
+    
+    We observe in this case that if we specify 1 repetition, the system instead
+    executes many times. Why? -- this is no longer the case??
```

Comment:
Yeah, I don't know what this comment means. How are we specifying one repetition in this case? I'll remove it since I'm not aware of any problems with the boards.